Configuration of the proxied HTTP request version.

ManickaP opened the pull request:
Adds configuration side to setting up `Version` and `VersioningPolicy` (for .NET 5 and higher) of an outbound HTTP Request.

Fixes #282

ManickaP commented on src/ReverseProxy/Middleware/ProxyInvokerMiddleware.cs:
```diff
@@ -80,6 +81,16 @@ public async Task Invoke(HttpContext context)
                 Transforms = routeConfig.Transforms,
             };
 
+            // Booooo is null in Invoke_Works test but reverseProxyFeature.ClusterConfig.HttpRequestOptions works, what is right?
```
This is causing the issue in the test.
If swapped for `reverseProxyFeature.ClusterConfig.HttpRequestOptions` it works. But I'm not sure what source of the value is correct in real life (outside of test mocks)???

ManickaP commented on test/ReverseProxy.Tests/Middleware/ProxyInvokerMiddlewareTests.cs:
```diff
@@ -78,7 +85,13 @@ public async Task Invoke_Works()
                     httpClient,
-                    It.IsAny<RequestProxyOptions>()))
+                    It.Is<RequestProxyOptions>(options => true
+// Booooo doesn't work
```
This is where it fails since the `cluster.Config` in `ProxyInvokerMiddleware.Invoke` is `null`.

ManickaP commented on test/ReverseProxy.Tests/Middleware/ProxyInvokerMiddlewareTests.cs:
```diff
@@ -99,6 +112,8 @@ public async Task Invoke_Works()
                 await task;
             }
+
+            // Booooo this gets stuck if Setup fails, maybe Mock<IHttpProxy>().Verify(); should be put before the await
```
And here it gets stuck due to the failure. It seems like `Returns` is never called and the task is never signaled to finish.

ManickaP commented on src/ReverseProxy/Configuration/ConfigurationConfigProvider.cs:
```diff
@@ -353,6 +354,22 @@ private ProxyHttpClientOptions Convert(ProxyHttpClientData data)
+
+            return new ProxyHttpRequestOptions
+            {
+                Version = Version.Parse(data.Version + (data.Version.Contains('.') ? "" : ".0")),
```
This is a hacky how to work around `Version` limitation of having at least major.minor in the string, see remarks of https://docs.microsoft.com/en-us/dotnet/api/system.version.parse?view=net-5.0

Tratcher replied:
This also needs to handle null or empty strings for data.Version.

ManickaP replied:
Fixed.

alnikola commented on src/ReverseProxy/Service/Management/ProxyConfigManager.cs:
```diff
@@ -311,6 +313,12 @@ private void UpdateRuntimeClusters(IList<Cluster> newClusters)
                                 httpClient,
                                 newClusterHttpClientOptions,
+                                new ClusterProxyHttpRequestOptions(
+                                    version: newCluster.HttpRequest?.Version ?? HttpVersion.Version20
```
The `Version` property type is an abstract `System.Version`, thus I think we need to ensure that the given value is a valid HTTP protocol version. Let's add a new validation method into [ConfigValidator](https://github.com/microsoft/reverse-proxy/blob/bb6978db0517b21d3125e65bcf8ebb6d87351fd6/src/ReverseProxy/Service/Config/ConfigValidator.cs#L109).

ManickaP replied:
Added.

alnikola commented on test/ReverseProxy.Tests/Service/HealthChecks/DefaultProbingRequestFactoryTests.cs:
```diff
@@ -31,7 +31,7 @@ public void CreateRequest_HealthEndpointIsNotDefined_UseDestinationAddress(strin
         {
             return new ClusterConfig(
-                new Cluster { Id = id }, new ClusterHealthCheckOptions(default, healthCheckOptions), default, default, null, default, null);
+                new Cluster { Id = id }, new ClusterHealthCheckOptions(default, healthCheckOptions), default, default, null, default, default, null);
```
If the request version settings are also applied to active health probes, it must be verified in this test.

ManickaP replied:
Added.

alnikola reviewed (changes requested):
Consider applying the request version settings to active health probes.

ManickaP commented on src/ReverseProxy/Middleware/ProxyInvokerMiddleware.cs:
```diff
@@ -80,6 +81,15 @@ public async Task Invoke(HttpContext context)
+            var requestOptions = reverseProxyFeature.ClusterConfig.HttpRequestOptions;
+            if (requestOptions.Version != default)
+            {
+                proxyOptions.Version = requestOptions.Version;
```
Duplicate of https://github.com/microsoft/reverse-proxy/pull/512#discussion_r517039131

ManickaP replied:
Added

ManickaP commented on src/ReverseProxy/Middleware/ProxyInvokerMiddleware.cs:
```diff
@@ -80,6 +81,15 @@ public async Task Invoke(HttpContext context)
                 Transforms = routeConfig.Transforms,
             };
 
+            var requestOptions = reverseProxyFeature.ClusterConfig.HttpRequestOptions;
```
Fixed.

alnikola commented on src/ReverseProxy/Service/Config/ConfigValidator.cs:
```diff
@@ -322,6 +324,23 @@ private void ValidateProxyHttpClient(IList<Exception> errors, Cluster cluster)
             }
         }
 
+        private void ValidateProxyHttpRequest(IList<Exception> errors, Cluster cluster)
```
Did you add a new `ConfigValidator` unit test covering this new logic? I saw only one for `ProxyConfigManager`, but we usually test `ConfigValidator` directly. Please, [check the existing tests](https://github.com/microsoft/reverse-proxy/blob/bb6978db0517b21d3125e65bcf8ebb6d87351fd6/test/ReverseProxy.Tests/Service/Config/ConfigValidatorTests.cs#L15).

ManickaP replied:
Added.

alnikola commented on test/ReverseProxy.Tests/Service/HealthChecks/DefaultProbingRequestFactoryTests.cs:
```diff
@@ -18,20 +21,31 @@ public class DefaultProbingRequestFactoryTests
         public void CreateRequest_HealthEndpointIsNotDefined_UseDestinationAddress(string address, string health, string healthPath, string expectedRequestUri)
         {
-            var clusterConfig = GetClusterConfig("cluster0", new ClusterActiveHealthCheckOptions(true, null, null, "policy", healthPath));
+            var version = HttpVersion.Version11;
```
Could please we make it a test parameter with at least two values '1.1' and 'null' to test [this branching](https://github.com/microsoft/reverse-proxy/blob/fca0d54e3a71cd4a710103cdbc5869967741a4f6/src/ReverseProxy/Service/HealthChecks/DefaultProbingRequestFactory.cs#L20)?

ManickaP replied:
No problem, but should I rather move the arguments into `[MemberData]` method? This will double the number of `[InlineData]` and there's already 6 lines of it.
@Tratcher once mentioned that complex [MemberData] impact test reporting and I haven't found single occurrence in the codebase. So are we against it in general? Or is this a good place to use it?

alnikola replied:
Actually, I was wrong above because it would be better to add a separate test since this is focused on URI handling. In there, you can set the fixed `Address` value to `https://localhost:10000/` and provide several combinations of `Version` and `VersionPolicy`. `[MemberData]` supplying `VersionPolicy` values can be put inside an `ifdef` block for unsupported platfroms.

ManickaP replied:
Added. 
cc @Tratcher: I used `MemberData` for this one, please check it. I'm thinking I could split the test even more and test `VersionPolicy` separately from the version.

alnikola reviewed (commented):
Production code looks good now, only a couple test-related suggestions.

ManickaP commented on src/ReverseProxy/Abstractions/ClusterDiscovery/Contract/ProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Microsoft.ReverseProxy.Abstractions
+{
+    public sealed class ProxyHttpRequestOptions
```
Added.

Tratcher commented on test/ReverseProxy.Tests/Service/HealthChecks/DefaultProbingRequestFactoryTests.cs:
```diff
@@ -18,20 +23,71 @@ public class DefaultProbingRequestFactoryTests
 
-        private ClusterConfig GetClusterConfig(string id, ClusterActiveHealthCheckOptions healthCheckOptions)
+        [Theory]
+        [MemberData(nameof(RequestVersionMemberData))]
```
Why does this test need to be exhaustive? The actual values don't affect the behavior of anything here. It should be sufficient to test that one non-default value can be set for both parameters. E.g. HttpVersion.Version10 and HttpVersionPolicy.RequestVersionExact.

ManickaP replied:
Fixed.

Tratcher commented on src/ReverseProxy/Service/Management/ProxyConfigManager.cs:
```diff
@@ -311,6 +313,13 @@ private void UpdateRuntimeClusters(IList<Cluster> newClusters)
                                 httpClient,
                                 newClusterHttpClientOptions,
+                                new ClusterProxyHttpRequestOptions(
+                                    requestTimeout: newCluster.HttpRequest?.RequestTimeout ?? TimeSpan.FromSeconds(100),
+                                    version: newCluster.HttpRequest?.Version ?? HttpVersion.Version20
+#if NET
+                                    , versionPolicy: newCluster.HttpRequest?.VersionPolicy ?? HttpVersionPolicy.RequestVersionOrLower
+#endif
```
Pass the nulls through, let the consumer choose the default if it's not set.
```suggestion
                                    requestTimeout: newCluster.HttpRequest?.RequestTimeout,
                                    version: newCluster.HttpRequest?.Version
#if NET
                                    , versionPolicy: newCluster.HttpRequest?.VersionPolicy
#endif
```

ManickaP replied:
fixed.

ManickaP commented on src/ReverseProxy/Abstractions/ClusterDiscovery/Contract/ProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,48 @@
+        public Version Version { get; set; }
+
+#if NET
+        public HttpVersionPolicy VersionPolicy { get; set; }
```
Fixed.

ManickaP commented on src/ReverseProxy/Service/HealthChecks/DefaultProbingRequestFactory.cs:
```diff
@@ -15,7 +16,18 @@ public HttpRequestMessage CreateRequest(ClusterConfig clusterConfig, Destination
+            }
+            return new HttpRequestMessage(HttpMethod.Get, probeUri)
+            {
+                Version = version,
```
Fixed.

ManickaP commented on src/ReverseProxy/Abstractions/ClusterDiscovery/Contract/ProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Microsoft.ReverseProxy.Abstractions
+{
+    public sealed class ProxyHttpRequestOptions
```
Added into the existing docs.

ManickaP commented:
@Tratcher, @alnikola: I think I've addressed all comments except for the [type-merging one](https://github.com/microsoft/reverse-proxy/pull/512#discussion_r517037962), which I'd rather do separately if you're OK with it. I'll file an issue and look into it immediately. I'd just prefer to split the work.

Tratcher reviewed (approved):
File an issue for the type merging.

davidfowl commented:
@ManickaP Hey, I'm working on this PR to change configuration reading https://github.com/microsoft/reverse-proxy/pull/521. I'm going to wait until this PR is merged.

ManickaP commented on src/ReverseProxy/Abstractions/ClusterDiscovery/Contract/ProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,64 @@
+        /// Timeout for the outgoing request.
+        /// Default is 100 seconds.
+        /// </summary>
+        public TimeSpan? RequestTimeout { get; set; }
```
Should I rename other places as well?
- https://github.com/microsoft/reverse-proxy/blob/0e1842357c349fac27bd5327f044d1508052576e/src/ReverseProxy/Service/RuntimeModel/ClusterProxyHttpRequestOptions.cs#L32
- https://github.com/microsoft/reverse-proxy/blob/5818b8ebdef44baaab54585a21e95d05d37d57f8/src/ReverseProxy/Configuration/Contract/ProxyHttpRequestData.cs#L15
and potentially:
- https://github.com/microsoft/reverse-proxy/blob/99b3d78121a20a0e56b2ecc9c69d8c768f908dff/src/ReverseProxy/Service/Proxy/RequestProxyOptions.cs#L25

Tratcher replied:
Fair point, nevermind.

ManickaP replied:
Reverted.

ManickaP commented on test/ReverseProxy.Tests/Service/Config/ConfigValidatorTests.cs:
```diff
@@ -650,21 +720,21 @@ public async Task EnablePassiveHealthCheck_Works(int? reactivationPeriod, string
+                            : (TimeSpan?)null
                     }
                 }
             };
```
Reverted, it was unintentional. Sorry.

ManickaP commented on src/ReverseProxy/Service/RuntimeModel/ClusterProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,49 @@
+    /// </summary>
+    public readonly struct ClusterProxyHttpRequestOptions
+    {
+        public ClusterProxyHttpRequestOptions(
```
Done.

ManickaP commented on docs/docfx/articles/proxyhttpclientconfig.md:
```diff
@@ -70,8 +75,25 @@ Configuration settings:
+
+Configuration settings:
+- RequestTimeout - the timeout for the outgoing request sent by [HttpMessageInvoker.SendAsync](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpmessageinvoker.sendasync?view=netcore-3.1). If not specified, 100 seconds is used.
+- Version - outgoing request [version](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httprequestmessage.version?view=netcore-3.1). The supported values at the moment are 1.0, 1.1 and 2. Default value is 2.
```
Done.

Tratcher commented on src/ReverseProxy/Service/RuntimeModel/ClusterProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,49 @@
+            Version = version;
+            VersionPolicy = versionPolicy;
+        }
+#else
+        public ClusterProxyHttpRequestOptions(TimeSpan? requestTimeout, Version version)
+        {
+            RequestTimeout = requestTimeout;
+            Version = version;
+        }
+#endif
```
This constructor should exist in both versions. We don't change/remove public API across versions if we can avoid it.
```suggestion
#endif
        public ClusterProxyHttpRequestOptions(TimeSpan? requestTimeout, Version version)
        {
            RequestTimeout = requestTimeout;
            Version = version;
        }

```

ManickaP replied:
Done.

ManickaP commented on src/ReverseProxy/Service/RuntimeModel/ClusterProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Microsoft.ReverseProxy.RuntimeModel
+{
+    public readonly struct ClusterProxyHttpRequestOptions
```
https://github.com/microsoft/reverse-proxy/issues/525

ManickaP commented:
@alnikola : could you please give this another look? I cannot merge till you approve since you originally requested changes.

alnikola reviewed (approved):
LGTM

Tratcher commented on src/ReverseProxy/Service/RuntimeModel/ClusterProxyHttpRequestOptions.cs:
```diff
@@ -0,0 +1,52 @@
+            RequestTimeout = requestTimeout;
+            Version = version;
+#if NET
+            VersionPolicy = HttpVersionPolicy.RequestVersionOrLower;
```
`VersionPolicy = null;` The default is applied later.

ManickaP replied:
Aaaa, you're right.
